app/Models/Mship/Concerns/HasRoles.php

```diff
@@ -0,0 +1,75 @@
+<?php
+
+namespace App\Models\Mship\Concerns;
+
+use App\Events\Mship\Roles\RoleAssigned;
+use App\Events\Mship\Roles\RoleRemoved;
+use Spatie\Permission\Traits\HasRoles as OriginalHasRoles;
+
+/**
+ * @mixin \App\Models\BaseModel
+ */
+trait HasRoles
+{
+    use OriginalHasRoles {
+        assignRole as protected originalAssignRole;
+        removeRole as protected originalRemoveRole;
+        syncRoles as protected originalSyncRoles;
+    }
+
+    /**
+     * @param mixed ...$roles
+     * @return $this
+     */
+    public function assignRole(...$roles)
+    {
+        $this->originalAssignRole(...$roles);
+
+        $this->fireRoleAssignedEvent($roles);
+
+        return $this;
+    }
+
+    /**
+     * @param $role
+     * @return bool
+     */
+    public function fireRoleAssignedEvent($role)
+    {
+        if (is_iterable($role)) {
+            return array_walk($role, [$this, 'fireRoleAssignedEvent']);
+        }
+
+        event(new RoleAssigned($this, $this->getStoredRole($role)));
+
+        return true;
+    }
+
+    /**
+     * @param $role
+     * @return $this
+     */
+    public function removeRole($role)
+    {
+        $this->originalRemoveRole($role);
+
+        $this->fireRoleRemovedEvent($role);
+
+        return $this;
+    }
+
+    /**
+     * @param $role
+     * @return bool
+     */
+    public function fireRoleRemovedEvent($role)
```

Comment:
There's a school of thought here that might say that passing an array of roles is simpler, even if it has only one element, and then just looping through the roles passed.

Reply:
We can't really. The package supports either (ref: https://spatie.be/docs/laravel-permission/v4/basic-usage/role-permissions);

```
$account->assignRole('role_one');
$account->assignRole('role_one', 'role_two');
$account->assignRole(['role_one', 'role_two']);
```

If I modified this method to only support arrays, the first two approaches wouldn't work.

I actually grabbed this code from one of the Spatie maintainers who had commented on a GH issue on the package asking for this to be added. Doesn't look like they planned to add it so they suggested this.